### PR TITLE
Do not collapse ssl_crtd requests across reconfigurations

### PR DESCRIPTION
**Pull request by eduard-bagdasaryan**
Squid collapses new ssl_crtd requests on a pending request with an
identical helper query. This collapsing “works” across Squid
reconfigurations even though the old helper responsible for the first
request is replaced with a new one during reconfiguration; that old
helper must complete servicing pending requests.

However, since the helper program itself could have been changed (just
prior to reconfiguration), it is conceptually wrong for
post-reconfiguration requests to reuse the old helper program response:
Squid must conservatively assume that the response may have changed
because the helper protocol does not allow Squid to validate the
freshness of the helper response (to a collapsed request). Such blind
reuse also creates runtime problems if a buggy helper never responds to
a request X, stalling all the new requests getting collapsed on X (until
a Squid restart).

The added squid.conf "tagging" or "versioning" scheme may be useful for
Squid DNS and HTTP collapsing features that lack freshness checks and,
hence, should not use potentially stale pending transactions as
collapsing targets for new transaction after a reconfiguration.

Also pool GeneratorRequests map entries.

### CODE REVIEW

File: src/tests/stub_SquidConfig.cc [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+
+#include "SquidConfig.h"
+
+InstanceIdDefinitions(SquidConfig, "config");
```
Comment (rousskov) [inline src/tests/stub_SquidConfig.cc:13]: If this code survives, let's keep it in sync with the other change request.

```suggestion
InstanceIdDefinitions(SquidConfig, "cfg");
```

File: src/Makefile.am [hunk 2]
```diff
@@ -817,6 +817,7 @@ tests_testClpMap_SOURCES = \
 nodist_tests_testClpMap_SOURCES = \
 	tests/stub_HelperChildConfig.cc \
 	tests/stub_SBuf.cc \
+	tests/stub_SquidConfig.cc \
```
Comment (rousskov) [inline src/Makefile.am:820]: Hmm... What happens if instead of linking with stub_SquidConfig.cc, we link with SquidConfig.cc itself? I know that the difference between the two files is that the latter also has Config and Config2 definitions. Do those definitions bring more linking problems? It feels odd that a stub file does not define those two globals (and if we add those definitions, this stub file will not be needed!)...

File: src/SquidConfig.h [hunk 3]
```diff
@@ -542,6 +550,12 @@ class SquidConfig
         int connect_gap;
         int connect_timeout;
     } happyEyeballs;
+
+    using Id = InstanceId<SquidConfig>;
+    /// A Squid process (re)configuration identifier.
+    /// Changes with each lifecycleStart() call even if
+    /// reconfiguration changed nothing (that Squid could detect).
```
Comment (rousskov) [inline src/SquidConfig.h:557]: Let's remove references to future features this code will eventually support and drop "re" prefix (because this data member identifies a given instance of Squid configuration).

```suggestion
    /// configuration ID that changes with each lifecycleStart() call
```

File: src/cache_cf.cc [hunk 4]
```diff
@@ -3918,6 +3920,20 @@ configFreeMemory(void)
 #if USE_OPENSSL
     Ssl::unloadSquidUntrusted();
 #endif
+    Config.lifecycleEnd();
+}
+
+void
+SquidConfig::lifecycleStart()
```
Comment (rousskov) [inline src/cache_cf.cc:3927]: Why are these SquidConfig methods defined here instead of Config.cc?

File: src/cache_cf.cc [hunk 5]
```diff
@@ -3918,6 +3920,20 @@ configFreeMemory(void)
...
+{
+    id.change();
+}
+
+void
+SquidConfig::lifecycleEnd()
+{
+    // if we decide to cache an SBuf representation of lifecycles_, then we can
+    // delete that object here
```
Comment (rousskov) [inline src/cache_cf.cc:3936]: Let's completely remove this effectively unused method. Its current body does not make much sense outside bag61. Its caller location is somewhat controversial. We will add it later if needed.

File: src/SquidConfig.cc [hunk 6]
```diff
@@ -9,6 +9,8 @@
 #include "squid.h"
 #include "SquidConfig.h"
 
+InstanceIdDefinitions(SquidConfig, "config");
```
Comment (rousskov) [inline src/SquidConfig.cc:12]: InstanceId prefixes tend to be short (because they are often printed as part of various call/transaction parameters, along with a lot of other information on the same line). I think we can shorten this to "cfg".

```suggestion
InstanceIdDefinitions(SquidConfig, "cfg");
```

**Review by rousskov — CHANGES_REQUESTED** [review]

File: src/SquidConfig.h [hunk 7]
```diff
@@ -542,6 +550,12 @@ class SquidConfig
...
+    Id id;
```
Comment (rousskov) [inline src/SquidConfig.h:558]: Please consider making this a private `id_` member so that we can protect it from accidental external modifications by returning a _constant_ reference to it.

**Review by eduard-bagdasaryan — COMMENTED** [review]
Done (abb4fe2).

File: src/ssl/helper.cc [hunk 8]
```diff
@@ -43,18 +43,34 @@ class GeneratorRequest {
...
     typedef std::vector<GeneratorRequestor> GeneratorRequestors;
     GeneratorRequestors requestors;
 };
 
-/// Ssl::Helper query:GeneratorRequest map
-typedef std::unordered_map<SBuf, GeneratorRequest*> GeneratorRequests;
+struct GeneratorRequestKeyHash
+{
+    std::size_t operator() (const GeneratorRequest::Key &pair) const {
+        return std::hash<decltype(pair.first)>()(pair.first) ^ std::hash<decltype(pair.second)>()(pair.second);
```
Comment (eduard-bagdasaryan) [inline src/ssl/helper.cc:64]: The compiler fails to deduce the template type without these explicit 'decltype' specifications.

File: src/tests/stub_SquidConfig.cc (hunk 1 above)
Comment (eduard-bagdasaryan) [reply src/tests/stub_SquidConfig.cc]: I removed this file.

File: src/ssl/helper.cc [hunk 9]
```diff
@@ -66,7 +82,7 @@ CBDATA_NAMESPACED_CLASS_INIT(Ssl, GeneratorRequest);
 static std::ostream &
 operator <<(std::ostream &os, const Ssl::GeneratorRequest &gr)
 {
-    return os << "crtGenRq" << gr.query.id.value << "/" << gr.requestors.size();
+    return os << "crtGenRq" << "/" << gr.key.first.id.detach() << "/" << gr.requestors.size();
```
Comment (eduard-bagdasaryan) [inline src/ssl/helper.cc:85]: Now with ScopedId the output will be: `HandleGeneratorReply: to 0x55d8f04d3740 in crtGenRq/SBuf12351/1`

Comment (rousskov) [reply src/ssl/helper.cc:85]: What I meant by ScopeId was something like

```suggestion
    return os << ScopedId("crtGenRq", gr.key.first.id.value) << '/' << gr.requestors.size();
```

Comment (eduard-bagdasaryan) [reply src/ssl/helper.cc]: Done (ce39608). I remembered your suggestion, but I noticed that the existing code does not call ScopedId() constructor with 2 parameters, except in InstanceId::detach(). I was thinking whether it is justifiable to separate InstanceId::value from its existing prefix in such a way.

Comment (rousskov) [reply src/ssl/helper.cc]: > [is it] justifiable to separate InstanceId::value from its existing prefix in such a way

Using one object ID to identify another object is a bit of a hack. It remains a bit of a hack _regardless_ of the implementation details. Unlike detach(), the ScopedId() implementation results in a much more readable ID that is consistent with other similar IDs.

----

If not banned (for no good reason), I am sure that this ID reuse will spread. We should encapsulate its correct implementation, especially given the number of iterations it took us to finalize one use case. I think that can be done by adding an `InstanceId::aka(const char *prefix)` "also known as" method that creates and returns a ScopedId object. I think that adding aka() should be done in a dedicated PR, to avoid attracting potentially unhelpful attention to this trivial piece of code and stalling the whole PR. If there are no similar reuses in the official code already, we should add aka() in a _followup_ PR.

**Comment by rousskov** [comment]
This became official PR 1469.
